File: src/NodeParser/StringTemplateLiteralNodeParser.ts
import ts from "typescript";
import { UnknownTypeError } from "../Error/UnknownTypeError.js";
import { Context, NodeParser } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { BaseType } from "../Type/BaseType.js";
import { LiteralType } from "../Type/LiteralType.js";
import { StringType } from "../Type/StringType.js";
import { UnionType } from "../Type/UnionType.js";
import { extractLiterals } from "../Utils/extractLiterals.js";

export class StringTemplateLiteralNodeParser implements SubNodeParser {
    public constructor(protected childNodeParser: NodeParser) {}

    public supportsNode(node: ts.NoSubstitutionTemplateLiteral | ts.TemplateLiteralTypeNode): boolean {
        return (
            node.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral || node.kind === ts.SyntaxKind.TemplateLiteralType
        );
    }
    public createType(node: ts.NoSubstitutionTemplateLiteral | ts.TemplateLiteralTypeNode, context: Context): BaseType {
        if (node.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral) {
            return new LiteralType(node.text);
        }

        try {
            const prefix = node.head.text;
            const matrix: string[][] = [[prefix]].concat(
                node.templateSpans.map((span) => {
                    const suffix = span.literal.text;
                    const type = this.childNodeParser.createType(span.type, context);
                    return extractLiterals(type).map((value) => value + suffix);
                })
            );

            const expandedLiterals = expand(matrix);

            const expandedTypes = expandedLiterals.map((literal) => new LiteralType(literal));

            if (expandedTypes.length === 1) {
                return expandedTypes[0];
            }

            return new UnionType(expandedTypes);
        } catch (error) {
            if (error instanceof UnknownTypeError) {
                return new StringType();
            }

            throw error;
        }
    }
}

function expand(matrix: string[][]): string[] {
    if (matrix.length === 1) {
        return matrix[0];
    }
    const head = matrix[0];
    const nested = expand(matrix.slice(1));
    const combined = head.map((prefix) => nested.map((suffix) => prefix + suffix));
    return ([] as string[]).concat(...combined);
}

File: internal/parser/string_template_literal_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// StringTemplateLiteralNodeParser handles no-substitution template literals
// and template literal types, expanding the cartesian product of the literal
// values of all spans (src/NodeParser/StringTemplateLiteralNodeParser.ts).
type StringTemplateLiteralNodeParser struct {
	childNodeParser NodeParser
}

func NewStringTemplateLiteralNodeParser(childNodeParser NodeParser) *StringTemplateLiteralNodeParser {
	return &StringTemplateLiteralNodeParser{childNodeParser: childNodeParser}
}

func (p *StringTemplateLiteralNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindNoSubstitutionTemplateLiteral || node.Kind == ast.KindTemplateLiteralType
}

func (p *StringTemplateLiteralNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	if node.Kind == ast.KindNoSubstitutionTemplateLiteral {
		return &types.LiteralType{Value: node.Text()}
	}

	template := node.AsTemplateLiteralTypeNode()
	prefix := template.Head.Text()
	matrix := [][]string{{prefix}}
	for _, span := range template.TemplateSpans.Nodes {
		typeSpan := span.AsTemplateLiteralTypeSpan()
		suffix := typeSpan.Literal.Text()
		typ := p.childNodeParser.CreateType(typeSpan.Type, ctx, nil)
		literals, ok := tryExtractLiterals(typ)
		if !ok {
			// A span whose type is not literal-expandable (e.g. `${string}`)
			// degrades the whole template to a plain string. This mirrors
			// catching UnknownTypeError in the TypeScript implementation.
			return &types.StringType{}
		}
		row := make([]string, len(literals))
		for i, value := range literals {
			row[i] = value + suffix
		}
		matrix = append(matrix, row)
	}

	expandedLiterals := expandTemplateMatrix(matrix)
	expandedTypes := make([]types.Type, len(expandedLiterals))
	for i, literal := range expandedLiterals {
		expandedTypes[i] = &types.LiteralType{Value: literal}
	}

	if len(expandedTypes) == 1 {
		return expandedTypes[0]
	}
	return types.NewUnionType(expandedTypes)
}

// tryExtractLiterals collects the literal expansions of a type, reporting
// failure instead of propagating the panic raised by types.ExtractLiterals
// on non-literal types.
func tryExtractLiterals(t types.Type) (literals []string, ok bool) {
	defer func() {
		if recover() != nil {
			literals = nil
			ok = false
		}
	}()
	return types.ExtractLiterals(t), true
}

// expandTemplateMatrix builds the cartesian product of the matrix rows,
// with the leftmost row varying slowest.
func expandTemplateMatrix(matrix [][]string) []string {
	if len(matrix) == 1 {
		return matrix[0]
	}
	head := matrix[0]
	nested := expandTemplateMatrix(matrix[1:])
	combined := make([]string, 0, len(head)*len(nested))
	for _, prefix := range head {
		for _, suffix := range nested {
			combined = append(combined, prefix+suffix)
		}
	}
	return combined
}
